### PR TITLE
Adds a method to exclude deleted batches from query

### PR DESCRIPTION
**Pull request by genypaul**
This PR adds a method to not take into account the batches that have been marked as `DELETED`.
It allows an app to be notified upon deletion right away.

### CODE REVIEW

**Comment by hal9002** [comment]
Can one of the admins review this PR?

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadProvider.java [hunk 1]
```diff
@@ -800,7 +800,9 @@ public int update(final Uri uri, final ContentValues values, final String where,
             case BATCHES_ID:
                 SqlSelection batchSelection = getWhereClause(uri, where, whereArgs, match);
                 count = db.update(DownloadContract.Batches.BATCHES_TABLE_NAME, values, batchSelection.getSelection(), batchSelection.getParameters());
-                notifyStatusIfBatchesStatusChanged(values);
+                if (values.containsKey(DownloadContract.Batches.COLUMN_STATUS) || values.containsKey(DownloadContract.Batches.COLUMN_DELETED)){
```
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadProvider.java]: That is what triggers the immediate notification :)

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadProvider.java [hunk 2]
```diff
@@ -520,7 +520,7 @@ private void notifyDownloadStatusChanged() {
         getContext().getContentResolver().notifyChange(downloadsUriProvider.getDownloadsWithoutProgressUri(), null);
     }
 
-    private void notifyBatchesStatusChanged() {
+    private void notifyBatchesChanged() {
```
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadProvider.java]: can we rename this method to make it explicit it excludes progress?

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchQuery.java [hunk 3]
```diff
@@ -15,6 +15,7 @@
     public static final BatchQuery ALL = new BatchQuery(null, null, null);
     private static final int LOW_END_FAILED_STATUS_CODE = 400;
     private static final int HIGH_END_FAILED_STATUS_CODE = 600;
+    private static final int BATCH_DELETED = 1;
```
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchQuery.java]: Could move this into DownloadsContract.Batches maybe?

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchQuery.java [hunk 4]
```diff
@@ -211,29 +221,45 @@ public Builder withExtraData(String extraData) {
...
+            setCriteriaListFrom(criteriaBuilders);
+
+            Criteria criteria = builder.build();
+            String selection = criteria.getSelection();
+            String sortOrder = criteria.getSort();
+            String[] selectionArguments = criteria.getSelectionArguments();
+            return new BatchQuery(selection, selectionArguments, sortOrder);
+        }
+
+        private void setCriteriaListFrom(List<Criteria.Builder> criteriaBuilders) {
```
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchQuery.java]: are there any tests around setting multiple criteria? Looks complex and would be good to make sure we don't break it :)

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchQuery.java (hunk 3 above)
Comment (genypaul) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchQuery.java]: It seems like a good place indeed :)

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchQuery.java (hunk 4 above)
Comment (genypaul) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchQuery.java]: There are some tests around query building in `BatchQueryTest`. But I did not see any with multiple criteria other than multiple ids + multiple statuses. I could maybe improve it to take into account other multiple criteria.

Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchQuery.java]: OK, well if those tests haven't broken (they haven't, the PR is still open!) then that's probably OK :)

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadProvider.java (hunk 2 above)
Comment (genypaul) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadProvider.java]: Actually. I'm more wondering why we only notify the `withoutProgress` uri and not the `allBatches` uri as well? Is there a reason

Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadProvider.java]: No you're right - we should notify both :+1:

Comment (genypaul) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadProvider.java]: https://github.com/novoda/download-manager/commit/b36b53ce71b6f4c28d307e1ee969832f97934199

**Comment by amlcurran** [comment]
LGTM :+1: but need more eyes

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchQuery.java [hunk 5]
```diff
@@ -211,29 +220,45 @@ public Builder withExtraData(String extraData) {
...
                     builder.and();
                 }
             }
+        }
 
+        private boolean isNotLast(Criteria.Builder criteriaBuilder, List<Criteria.Builder> criteriaBuilders) {
+            return criteriaBuilders.indexOf(criteriaBuilder) != (criteriaBuilders.size() - 1);
+        }
+
+        private List<Criteria.Builder> setCriteriaBuilders() {
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchQuery.java]: `combineCriteriaBuilders` ?

**Comment by blundell** [comment]
:+1:

File: library/src/main/java/com/novoda/downloadmanager/lib/BatchQuery.java (hunk 5 above)
Comment (genypaul) [inline library/src/main/java/com/novoda/downloadmanager/lib/BatchQuery.java]: Better. Too much `set*` and `build*` :smirk:
https://github.com/novoda/download-manager/commit/7f71c4f5c16578217f2a5382649e105575fa71cd

**Pull request merged**